velox/expression/tests/ArgGenerator.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/fuzzer/Utils.h"
+
+namespace facebook::velox::test {
+
+class ArgGenerator {
+ public:
+  virtual ~ArgGenerator() = default;
+
+  /// Given a signature and a concrete return type return randomly selected
+  /// valid input types. Returns empty vector if no input types can produce the
+  /// specified result type.
+  virtual std::vector<TypePtr> generateArgs(
+      const exec::FunctionSignature& signature,
+      const TypePtr& returnType,
+      FuzzerGenerator& rng) = 0;
+};
+
+/// An argument type generator for decimal function. A map keyed on the pair of
+/// precision and scale could be initialized with all possible input types.
+/// Argument types are generated by looking up the map with the precision and
+/// scale of return type, and randomly selecting valid input types.
+class DecimalArgGeneratorBase : public ArgGenerator {
```

Comment:
In the test, I'm seeing that you create derived classes for single-argument and two-argument functions. I assume these are common and wonder if it makes sense to have these classes available in production code.

For example, we can modify initialize method here to take the number of decimal arguments as input. Then derived classes can simply call initialize(1) or initialize(2) and provide matching toReturnType implementation. And we can make only initialize() and toReturnType() protected and everything else private. What do you think?

Reply:
Thanks for the suggestion. Updated.